Use Python tool for get_extrnl_mdl_file tasks

christinaholtNOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
These changes hook in the Python-based data ingest tool, replacing the previous scripts that handled this work as part of the get_extrn_mdl_file task. No attempt was made in this PR to replace the NOMADS fetching script with the Python utility, but the NOMADS data location has been added to the data_locations.yml file.

The functionality to write the data summary file has also been added to the Python tool to match the capabilities of the existing workflow tools.

I "reviewed" the PR to leave comments where it might be helpful for reviewing.

## TESTS CONDUCTED: 
I ran all the 25km tests in grid_* and nco_* on Hera and was successfully able to stage data from disk as previously expected. Several jobs failed in the same way as had been previously documented.

On Jet, I tested the same set of 25 km tests using the sparse data staged on disk to test the alternate data_locations available on that platform (known data streams for Jet, HPSS, BDP, and NOMADS). These tests completed as expected -- some ran to completion by finding the data they needed on HPSS or AWS, while others failed to find any data and returned a correct non-zero exit status.

I pushed a couple of additional changes to retrieve_data.py as the "alternate data locations" logic was tested and needed minor modification.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -924,7 +924,9 @@ machine (MACHINE):
-      EXTRN_MDL_SOURCE_BASEDIR_ICS="${EXTRN_MDL_SOURCE_BASEDIR_ICS}/${FV3GFS_FILE_FMT_ICS}"
+      EXTRN_MDL_SOURCE_BASEDIR_ICS="${EXTRN_MDL_SOURCE_BASEDIR_ICS}/${FV3GFS_FILE_FMT_ICS}/\${DATE_FIRST_CYCL}\${CYCL_HRS[0]}"
+    else
+      EXTRN_MDL_SOURCE_BASEDIR_ICS="${EXTRN_MDL_SOURCE_BASEDIR_ICS}/\${DATE_FIRST_CYCL}\${CYCL_HRS[0]}"
```
The retrieve_data.py script assumes nothing about the directory structure of the staged data, so the entire path must be templated here. The same is true below for the LBCS.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.sh:
```diff
@@ -958,32 +961,26 @@ boundary conditions specification interval (LBC_SPEC_INTVL_HRS):
       if [ "${FV3GFS_FILE_FMT_LBCS}" = "nemsio" ]; then
-        EXTRN_MDL_FILES_LBCS=( "${EXTRN_MDL_FILES_LBCS[@]/#/gfs.atmf}" )
-        EXTRN_MDL_FILES_LBCS=( "${EXTRN_MDL_FILES_LBCS[@]/%/.nemsio}" )
+        EXTRN_MDL_FILES_LBCS=( 'gfs.atmf{fcst_hr:03d}.nemsio' )
```
Here, we only need the template of the file(s) to be retrieved. More than one can be added in this array. The python script loops through the necessary forecast hours.

christinaholtNOAA commented on scripts/exregional_get_extrn_mdl_files.sh:
```diff
@@ -80,715 +72,89 @@ process_args valid_args "$@"
+  --debug \
+  --anl_or_fcst ${anl_or_fcst} \
+  --config ${USHDIR}/templates/data_locations.yml \
+  --cycle_date ${extrn_mdl_cdate} \
```
Note that all the templating done with dates in the Python script are done on the cycle of the external model. The script has no knowledge of the offset from the current cycle of SRW since that information is irrelevant to the process of grabbing the required data.

christinaholtNOAA commented on ush/config_defaults.sh:
```diff
@@ -711,10 +703,10 @@ EXTRN_MDL_SYSBASEDIR_LBCS=''
+EXTRN_MDL_SOURCE_BASEDIR_ICS=""
+EXTRN_MDL_FILES_ICS=""
+EXTRN_MDL_SOURCE_BASEDIR_LBCS=""
+EXTRN_MDL_FILES_LBCS=""
```
If these are empty by default, it becomes easier to manage their values when provided.

JeffBeck-NOAA replied:
It's good to have placeholder defaults so that users understand the format to employ.  Can examples be provided in the above comment block?

JeffBeck-NOAA replied:
Also, just to follow up on our email discussion, will this Python script be able to handled staged data for multiple cycles?  That was the previous functionality and the SRW App was set up to allow users to configure and run many cycles at once.

JeffBeck-NOAA replied:
@christinaholtNOAA, just pinging you on these questions.  Thanks!

christinaholtNOAA replied:
I can include the example in the comments, but if there is no good default that matches most use cases, then I think it makes sense to leave it as an empty string.

JeffBeck-NOAA replied:
Yeah, I think an empty string is fine, but an example in the comments would be great.  Also, will this Python script be able to handle staged data for multiple cycles?  Thanks.

christinaholtNOAA replied:
It will not handle multiple cycles. I haven't seen a use case for that functionality in the workflow. Rocoto and scripts handle multiple cycles, and this will be a tool that is called per Rocoto-cycle just like the previous cycling strategy.

JeffBeck-NOAA replied:
The SRW App currently allows users to choose a DATE_FIRST_CYCL and DATE_LAST_CYCL in config.sh to run multiple cycles at once through the Rocoto XML.  The get_extrn_ics and get_extrn_lbcs tasks grab external model data for all cycles from HPSS or from on disk at once.  If the Python script doesn't handle multiple cycles, is it looped for each of the cycles that a user sets between DATE_FIRST_CYCLE and DATE_LAST_CYCL?

christinaholtNOAA replied:
The `get_extrn_ics` and `get_extrn_lbcs` tasks run for each forecast cycle (as opposed to the at_start cycle) defined by the cycledef in Rocoto, and only need to handle data gathering for that cycle. To my knowledge, there are no multi-cycle tests in the WE2E tests that test this feature.

As I mentioned in the description, I only attempted to replace the data ingest features in the `get_extrn_ics` and `get_extrn_lbcs` tasks and not the NOMADS feature that gathers all data when `generate_FV3LAM_wflow.sh` is run. That feature has not worked for me in the past when I run those tests, nor when I use the Python script to pull the data. More debugging is needed for NOMADS data, and it should only be run as an optional stand-alone task for users on machines where compute nodes don't have access to the internet. It will absolutely not work for real time applications, or for longer retrospectives that need to more carefully manage disk space.

JeffBeck-NOAA replied:
The NOMADS data sourcing option should be considered secondary to HPSS and/or data stored on disk by the user, as their servers are often down or extremely slow.  It's completely fine not to worry about that here.

We should definitely add a multi-cycle WE2E test, but I'll go ahead and run manual test of that with this code today.

christinaholtNOAA commented on ush/retrieve_data.py:
```diff
@@ -43,7 +46,7 @@ def clean_up_output_dir(expected_subdir, local_archive, output_path, source_path
     # Check to make sure the files exist on disk
     for file_path in source_paths:
-        local_file_path = os.path.join(output_path,file_path)
+        local_file_path = os.path.join(output_path, file_path.lstrip("/"))
```
TIL: You can't join paths with os.path.join() when the second one starts with a slash!

christinaholtNOAA commented on ush/retrieve_data.py:
```diff
@@ -310,8 +317,9 @@ def get_requested_files(cla, file_names, input_loc, method='disk'):
-            loc = os.path.join(input_loc, file_name)
+        for file_template in file_templates:
+            loc = os.path.join(input_loc, file_template)
+            logging.debug(f'Full file path: {loc}')
```
I changed the behavior here slightly to make it easier to call from bash. The file_template is now just a list, whereas before it was meant to be the same YAML structure as in the data_locations.yml file. That was totally a pain, and unnecessary.

christinaholtNOAA commented on ush/retrieve_data.py:
```diff
@@ -505,6 +519,30 @@ def setup_logging(debug=False):
+    with open(summary_fp, "w") as summary:
+        summary.write(file_contents)
+
+
```
This is very specific to the SRW, and I'm not a huge fan of that, but am fine with it for now. Let me know if this is problematic.

christinaholtNOAA commented on ush/retrieve_data.py:
```diff
@@ -521,29 +559,31 @@ def main(cla):
+        msg = dedent(f'''No data stores have been defined for
+               {cla.external_model}!''')
+        if cla.input_file_path is None:
+            data_stores = ['disk']
```
This allows us to provide disk locations for data that may not have a defined known location.

christinaholtNOAA commented on ush/retrieve_data.py:
```diff
@@ -672,9 +714,13 @@ def parse_args():
         help='A path to data stored on disk. The path may contain \
         Python templates. File names may be supplied using the \
-        --file_names flag, or the default naming convention will be \
+        --file_templates flag, or the default naming convention will be \
```
This was a more descriptive variable name, so I went ahead and made the shift as I hooked it into the scripts..

christinaholtNOAA commented on ush/templates/FV3LAM_wflow.xml:
```diff
@@ -281,7 +281,6 @@ MODULES_RUN_TASK_FP script.
     <envar><name>PDY</name><value><cyclestr>@Y@m@d</cyclestr></value></envar>
     <envar><name>CDATE</name><value><cyclestr>@Y@m@d@H</cyclestr></value></envar>
     <envar><name>CYCLE_DIR</name><value><cyclestr>&CYCLE_BASEDIR;/@Y@m@d@H</cyclestr></value></envar>
-    <envar><name>EXTRN_MDL_NAME</name><value>{{ extrn_mdl_name_ics }}</value></envar>
```
This information is unnecessary since it's included in the var_defns.sh file.

JeffBeck-NOAA commented on scripts/exregional_get_extrn_mdl_files.sh:
```diff
@@ -80,715 +72,89 @@ process_args valid_args "$@"
-num_files_found_on_disk="0"
-min_age="5"  # Minimum file age, in minutes.
 
-for fp in "${extrn_mdl_fps_on_disk[@]}"; do
```
When the real-time data files are either created on Jet, or transferred from WCOSS, I'm pretty sure they populate chunk by chunk, so there is still the chance of the script grabbing a file that is incomplete.  For staged data, I agree, that shouldn't be an issue.

JeffBeck-NOAA replied:
@christinaholtNOAA, pinging you on this question.  Thanks!

christinaholtNOAA replied:
I can work on adding it back if it's necessary.

JeffBeck-NOAA replied:
I think it's important to have for users who will be sourcing real-time data that are copied over from WCOSS to other NOAA HPC.

christinaholtNOAA replied:
@JeffBeck-NOAA Do we want this script to handle that, or Rocoto via dependency? Does it make sense to start a task before the data is ready?

christopherwharrop-noaa replied:
@JeffBeck-NOAA - You should never use a time dependency to signal data availability.  That is an anti-pattern that is guaranteed to fail at some point.  Always use a data dependency on the files you are looking for with an appropriate "age" attribute to account for files that get populated over a period of time.

christopherwharrop-noaa replied:
@christinaholtNOAA - I don't think it is ever safe to make assumptions about how data is populated via various delivery services.  If you're wrong, tasks will get triggered before data is truly available and may fail depending on the race between how quickly the job starts to run and how quickly the data is populated.  I think it's better to have some assurance the data really is available.  Checking how long ago the file was last modified is one way to do it.  There may be other approaches that are as good or better.

christinaholtNOAA replied:
@christopherwharrop-noaa Do you prefer a data dependency or a check in this script? I was leaning toward a datadependency since there is currently no dependency on the get_extrn_* tasks now. The current behavior of the workflow in no way supports near-real time runs (retro-only is hard coded in XML template, and there are no dependencies on getting the data), so I was assuming that wouldn't be a feature I'd need to support with this refactor. 

If we do intend to support real-time runs, many new features will be needed for that. I'm happy to add a datadep as part of this work, and to add a second checker inside the Python tool if we feel those are necessary now. I tend to go for "build it when it's needed", though.

christinaholtNOAA commented:
Can someone please review this PR?

JeffBeck-NOAA commented:
> Can someone please review this PR?

Hi @christinaholtNOAA, I reviewed this PR a while ago, then pinged you on my comments three days ago.  I guess you're not getting the GitHub emails on those comments?  Weird!

JeffBeck-NOAA reviewed (commented):
Please see comments, thanks!

christinaholtNOAA commented:
I guess I am not, or they are getting mixed in with all the others. Thanks for the review @JeffBeck-NOAA! I will address those comments soon.

christinaholtNOAA commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
 #-----------------------------------------------------------------------
 #
-case ${EXTRN_MDL_NAME} in
+
```
Wow! At least I was consistent. Gee! I fixed those locally, and will push after I address subsequent comments.

gsketefian replied:
Ok, thanks!

JeffBeck-NOAA commented:
@christinaholtNOAA, I tested this PR on Hera with the standard FV3GFS external model data retrieval from HPSS for two cycles: 2019061500 and 2019061600 using the 25-km CONUS domain, but received the following error:

INFO: Checking disk for FV3GFS

DEBUG: User supplied file names are: {'grib2': {'anl': ['gfs.t{hh}z.pgrb2.0p25.f000'], 'fcst': ['gfs.t{hh}z.pgrb2.0p25.f{fcst_hr:03d}']}, 'nemsio': {'anl': ['gfs.t{hh}z.atmanl.nemsio', 'gfs.t{hh}z.sfcanl.nemsio'], 'fcst': ['gfs.t{hh}z.atmf{fcst_hr:03d}.nemsio', 'gfs.t{hh}z.sfcf{fcst_hr:03d}.nemsio']}}

INFO: Getting files named like {'grib2': {'anl': ['gfs.t{hh}z.pgrb2.0p25.f000'], 'fcst': ['gfs.t{hh}z.pgrb2.0p25.f{fcst_hr:03d}']}, 'nemsio': {'anl': ['gfs.t{hh}z.atmanl.nemsio', 'gfs.t{hh}z.sfcanl.nemsio'], 'fcst': ['gfs.t{hh}z.atmf{fcst_hr:03d}.nemsio', 'gfs.t{hh}z.sfcf{fcst_hr:03d}.nemsio']}}

INFO: Retrieved files will be placed here:
 /scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/Christina/2019061500/FV3GFS/for_ICS

Traceback (most recent call last):
  File "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py", line 716, in <module>
    main(CLA)
  File "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py", line 533, in main
    unavailable = get_requested_files(cla,
  File "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py", line 275, in get_requested_files
    loc = os.path.join(input_loc, file_template)
  File "/contrib/miniconda3/4.5.12/envs/regional_workflow/lib/python3.8/posixpath.py", line 90, in join
    genericpath._check_arg_types('join', a, *p)
  File "/contrib/miniconda3/4.5.12/envs/regional_workflow/lib/python3.8/genericpath.py", line 152, in _check_arg_types
    raise TypeError('%s() argument must be str or bytes, not %r' %
TypeError: join() argument must be str or bytes, not 'dict'

ERROR:
  From script:  "exregional_get_extrn_mdl_files.sh"
  Full path to script:  "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/scripts/exregional_get_extrn_mdl_files.sh"
Call to retrieve_data.py failed with a non-zero exit status.

The command was:

python3 -u /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py   --debug   --anl_or_fcst anl   --config /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/templates/data_locations.yml   --cycle_date 2019061500   --data_stores disk hpss aws   --external_model FV3GFS   --fcst_hrs 0   --output_path /scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/Christina/2019061500/FV3GFS/for_ICS   --input_file_path /scratch1/NCEPDEV/rstprod/com/gfs/prod/gfs.${yyyymmdd}/${hh}/atmos   --summary_file extrn_mdl_var_defns.sh      --file_type grib2

Exiting with nonzero status.

christinaholtNOAA commented:
@JeffBeck-NOAA Interesting catch. I will make the change that should fix your failure mode.

christinaholtNOAA commented:
@JeffBeck-NOAA Would you give your test a try again after updating the script. Resubmitting the job via rocoto should be sufficient. No need to re-configure the experiment or anything.

JeffBeck-NOAA commented:
> @JeffBeck-NOAA Would you give your test a try again after updating the script. Resubmitting the job via rocoto should be sufficient. No need to re-configure the experiment or anything.

Thanks, @christinaholtNOAA.  I reran things after pull your latest commits, but I'm still getting the same error, unfortunately.

christinaholtNOAA commented:
Thanks for trying again, @JeffBeck-NOAA. I'll need to take a closer look at this use case.

JeffBeck-NOAA commented:
Thanks, @christinaholtNOAA.  I'm happy to test again when you believe this issue is resolved.

christinaholtNOAA commented:
@JeffBeck-NOAA I think I've possibly taken care of it. I was able to run the task for an NCO test, which I believe sets a path, but no file names.

JeffBeck-NOAA commented:
@christinaholtNOAA, it's now working for FV3GFS data!  I also tested with HRRR and RAP data, but I'm getting the following error below.  Also, I see that for --external_model there are both HRRR/RAP and HRRRX/RAPX options.  We've had some internal DTC conversations on moving to support only the (past and current) operational HRRR and RAP data, since the experimental versions of HRRR and RAP are not longer running at GSL.  We had gotten the HRRRX and RAPX working in the SRW App at the time to support the latest GSL code, but I think now that the final HRRR and RAP implementations into operations are complete, we can remove that support.

usage: retrieve_data.py [-h] --anl_or_fcst {anl,fcst} [--config CONFIG]
                        --cycle_date CYCLE_DATE --data_stores
                        [DATA_STORES [DATA_STORES ...]] --external_model
                        {FV3GFS,GSMGFS,HRRR,NAM,RAP,RAPx,HRRRx} --fcst_hrs
                        FCST_HRS [FCST_HRS ...] --output_path OUTPUT_PATH
                        [--debug]
                        [--file_templates [FILE_TEMPLATES [FILE_TEMPLATES ...]]]
                        [--file_type {grib2,nemsio,netcdf}]
                        [--input_file_path INPUT_FILE_PATH]
                        [--summary_file SUMMARY_FILE]
retrieve_data.py: error: argument --input_file_path: expected one argument

ERROR:
  From script:  "exregional_get_extrn_mdl_files.sh"
  Full path to script:  "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/scripts/exregional_get_extrn_mdl_files.sh"
Call to retrieve_data.py failed with a non-zero exit status.

The command was:

python3 -u /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py   --debug   --anl_or_fcst anl   --config /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/templates/data_locations.yml   --cycle_date 2019061500   --data_stores disk hpss aws   --external_model HRRR   --fcst_hrs 0   --output_path /scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/Christina_HRRR_RAP/2019061500/HRRR/for_ICS   --input_file_path    --summary_file extrn_mdl_var_defns.sh

Exiting with nonzero status.

christinaholtNOAA commented:
@JeffBeck-NOAA Are you running cases from the WE2E suite?

JeffBeck-NOAA commented:
@christinaholtNOAA, I'm running what would be the equivalent of this WE2E: config.grid_RRFS_CONUS_3km_ics_HRRR_lbcs_RAP_suite_GFS_v15p2.sh.

christinaholtNOAA commented:
@JeffBeck-NOAA Let me re-run the test suite. I didn't think I'd made changes since I opened the PR, but something isn't right. Thanks for all this iteration!

JeffBeck-NOAA commented:
@christinaholtNOAA, sounds good. Thanks, and happy to help!

gsketefian commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
+#
+#-----------------------------------------------------------------------
+#
+if [ "${ICS_OR_LBCS}" = "ICS" ]; then
```
The nested if-statement below seems like a left-over from the original file (or copied over) that you don't need.  Just this:
```
time_offset_hrs=${EXTRN_MDL_ICS_OFFSET_HRS:-0}
```
is all you need I think.

gsketefian commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
+elif [ "${ICS_OR_LBCS}" = "LBCS" ]; then
+  time_offset_hrs=${EXTRN_MDL_LBCS_OFFSET_HRS:-0}
+  extrn_mdl_name=${EXTRN_MDL_NAME_LBCS}
+fi
```
Ok, thanks for fixing that up.

gsketefian commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
+ble for the specified cycle date and time (extrn_mdl_cdate) because the latter is
+earlier than the implementation date of this model:
+  extrn_mdl_name = \"${name}\"
+  CDATE_min = \"${end_date}\"
```
`CDATE_min` here should be lower case?  Or `cdate_max` above upper case?  Not sure...

gsketefian commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
 
 "RAP")
 # Examination of the HPSS archives shows that the RAPX data goes back to
 # July 01, 2015.
```
I'm assuming we don't want to have any references to RAPX (or HRRRX) any more (in the comment).  

Also, I can't remember how we decided on the `CDATE_min="2015070100"`.  @JeffBeck-NOAA Do you remember?  [This info](https://rapidrefresh.noaa.gov/) shows the RAP going further back.

JeffBeck-NOAA replied:
@gsketefian, we must have gone back in time to see what the earliest available data were on HPSS for RAPX.  Since we're moving to support only operational RAP data, I would think we could remove this comment.

gsketefian commented on modulefiles/tasks/cheyenne/get_extrn_ics:
```diff
@@ -1,5 +1,4 @@
-#############################################################
+#%Module
 
+module load pylib_regional_workflow
```
Nice.  Makes things much more compact.

gsketefian commented on modulefiles/tasks/hera/make_ics.local:
```diff
@@ -1,5 +1,4 @@
 #%Module
-module use -a /contrib/miniconda3/modulefiles
-module load miniconda3
+module load wgrib2/2.0.8 
```
I'm curious how this task was working before without `wgrib2`.

JeffBeck-NOAA replied:
@gsketefian, it's in the srw_common module file, loaded by the build environment file, which is loaded for each task.

gsketefian commented on ush/templates/data_locations.yml:
```diff
@@ -103,6 +103,28 @@ FV3GFS:
     file_names:
       <<: *gfs_file_names
 
+GSMGFS:
```
I noticed RAPx and HRRRx are still in this file.  Should they be removed?

JeffBeck-NOAA replied:
Yeah, in another comment, I mentioned that we'll only be supported the operational RAP and HRRR data moving forward.

gsketefian commented on jobs/JREGIONAL_GET_EXTRN_MDL_FILES:
```diff
@@ -72,66 +63,134 @@ print_info_msg "
-directory (either from disk or HPSS) the external model files from which
-initial or boundary condition files for the FV3 will be generated.
+This is the J-job script for the task that copies/fetches to a local
+directory (either from disk, HPSS, or URL) the external model files from
```
@christinaholtNOAA I think you said in one of the comments in the python data ingest script that the script doesn't currently support URLs.  So maybe remove that from this comment (unless I misunderstood)?  But if it is supported, remove the "either"!

christinaholtNOAA commented:
I think I've either addressed all the comments by changing code, or by replying to the questions. I've rerun the 25 km community and nco tests on Hera, and they have passed as much as is expected. 

@JeffBeck-NOAA Fingers crossed your multi-cycle test would now pass.

JeffBeck-NOAA commented:
Thanks, @christinaholtNOAA!  I just tried it again on FV3GFS and RAP/HRRR data.  The former still works, but the latter produces the error below.  Is this related to the external model file location set in the machine files?  If so, one solution might be to allow the input_file_path to be empty/undefined, and if so, skip from checking on disk to pulling from HPSS.

Traceback (most recent call last):
  File "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py", line 703, in <module>
    raise argparse.ArgumentTypeError(
argparse.ArgumentTypeError: You must provide an input_file_path when choosing  disk as a data store!
+ print_err_msg_exit 'Call to retrieve_data.py failed with a non-zero exit status.

The command was:

python3 -u /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/retrieve_data.py   --debug   --anl_or_fcst anl   --config /scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/ush/templates/data_locations.yml   --cycle_date 2019061500   --data_stores disk hpss aws   --external_model HRRR   --fcst_hrs 0   --output_path /scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/Christina_HRRR_RAP/2019061500/HRRR/for_ICS   --summary_file extrn_mdl_var_defns.sh
'

ERROR:
  From script:  "exregional_get_extrn_mdl_files.sh"
  Full path to script:  "/scratch2/BMC/fv3lam/beck/FV3-LAM/ufs-srweather-app_C/regional_workflow/scripts/exregional_get_extrn_mdl_files.sh"
Call to retrieve_data.py failed with a non-zero exit status.

christinaholtNOAA commented:
@JeffBeck-NOAA I am working on this case that is failing for you. I can definitely fix that case. I am curious if this case is from the WE2E tests. I need to incorporate a test like this in my standard suite, if so.

I need to resolve some conflicts, so will need to run the suite again after I make this change.

christinaholtNOAA commented:
@JeffBeck-NOAA Are those not HRRRx and RAPx paths?

jwolff-ncar commented:
Just FYI (may not be helpful or cause confusion) Here are the paths I have used for in the past. I don't know how they differ from what Jeff sent.
HRRR: /NCEPPROD/hpssprod/runhistory/rh${YYYY}/${YYYY}${MM}/${YYYY}${MM}${DD}/com_hrrr_prod_hrrr.${YYYY}${MM}${DD}_conus${SHH}-${EHH}.wrf.tar
RAP: /NCEPPROD/hpssprod/runhistory/rh${YYYY}/${YYYY}${MM}/${YYYY}${MM}${DD}/com_rap_prod_rap.${YYYY}${MM}${DD}${SHH}-${EHH}.wrf.tar
RAP (wrfnat): /NCEPPROD/hpssprod/runhistory/rh${YYYY}/${YYYY}${MM}/${YYYY}${MM}${DD}/gpfs_hps_nco_ops_com_rap_prod_rap.${YYYY}${MM}${DD}${SHH}-${EHH}.wrf.tar

JeffBeck-NOAA commented:
@christinaholtNOAA, sorry, there was some confusion on the HPSS Google Doc that I was referencing.  It looks like these paths will work and have data for 20190615:

RAP:

/NCEPPROD/hpssprod/runhistory/rhYYYY/YYYYMM/YYYYMMDD/*rap*wrf.tar

HRRR:

/NCEPPROD/hpssprod/runhistory/rhYYYY/YYYYMM/YYYYMMDD/*hrrr*conus*wrf.tar

JeffBeck-NOAA commented:
Thanks, @jwolff-ncar, those are the same paths as mine.  It looks like we should use the wrfprs files for RAP, since that wrfnat path is no longer valid after/during 2019 at some point.

christinaholtNOAA commented:
@JeffBeck-NOAA There are a couple of issues. 

First, I am seeing files that the RAP files exist in that location under a different name. I can add that alternative template to look for, if you think that is necessary. The files on this date are named like this: gpfs_hps_nco_ops_com_rap_prod_rap.2019061412-17.wrf.tar 

Second, the make_ics tasks fail. I am assuming because of the older version of HRRR.

If we are dropping support for anything other than operational RAP/HRRR, should we maybe be testing with a newer test case? I believe the failures that are showing up for this case are because of unsupported input instead of problems with the new data ingest method.

JeffBeck-NOAA commented:
@christinaholtNOAA 

> First, I am seeing files that the RAP files exist in that location under a different name. I can add that alternative template to look for, if you think that is necessary. The files on this date are named like this: gpfs_hps_nco_ops_com_rap_prod_rap.2019061412-17.wrf.tar

Yeah, the RAP wrfnat file names apparently changes at some point between 2019 and 2021.  The best option is to have the staging script check for both templates:

`gpfs_hps_nco_ops_com_rap_prod_rap.YYYYMMDDHH-HH.wrf.tar`

and

`com_rap_prod_rap.YYYYMMDDHH-HH.wrf.tar`

in the following location:

`/NCEPPROD/hpssprod/runhistory/rhYYYY/YYYYMM/YYYYMMDD`

> Second, the make_ics tasks fail. I am assuming because of the older version of HRRR.

I was able to get the HRRR wrfprs data from the following location to run without an issue (for 2019061500 and 2021061500)

`/NCEPPROD/hpssprod/runhistory/rhYYYY/YYYYMM/YYYYMMDD/com_hrrr_prod_hrrr.YYYYMMDD_conusHH-HH.wrf.tar`

Here are my two runs on Hera (I stopped them after chgres_cube completed):

```
/scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/HRRR_RAP_2019061500
/scratch2/BMC/fv3lam/beck/FV3-LAM/expt_dirs/HRRR_RAP_2021061500
```

> If we are dropping support for anything other than operational RAP/HRRR, should we maybe be testing with a newer test case? I believe the failures that are showing up for this case are because of unsupported input instead of problems with the new data ingest method.

While a case from 2019 isn't too bad, it may be worth getting a newer date in the future to test the current operational configuration.  I'm not sure what failure you ran into, but I agree, it's not from the new data ingest method.

christinaholtNOAA commented:
I am running a test with the paths you provided, and will add those to the RAP/HRRR sections of the data_locations.yaml. And I have resolved the conflicts as of late last week.

If that works, what else still needs to be done? I'm not sure where this stands.

JeffBeck-NOAA commented:
> I am running a test with the paths you provided, and will add those to the RAP/HRRR sections of the data_locations.yaml. And I have resolved the conflicts as of late last week.
> 
> If that works, what else still needs to be done? I'm not sure where this stands.

I think that will be it!  I'll run a final test, then approve, and we can get this merged.

I did want to ask about sourcing external model data from online locations.  I saw that AWS will be supported, which is excellent!   Were there plans to expand this to [other data sources](https://dtcenter.org/nwp-containers-online-tutorial/publicly-available-data-sets) later?  This could be leveraged in a future PR, although some sources may be redundant with what is available on AWS.  I honestly am not sure we should invest more time in the previous NOMADS data staging feature, given it's propensity to be either down or extremely slow.

JeffBeck-NOAA commented:
@christinaholtNOAA, yes, I staged HRRR wrfprs files manually from this location on HPSS:

/NCEPPROD/hpssprod/runhistory/rhYYYY/YYYYMM/YYYYMMDD/com_hrrr_prod_hrrr.YYYYMMDD_conusHH-HH.wrf.tar

What files are you using for HRRR data and from where?

christinaholtNOAA commented:
Okay, so you are using wrfprs files, not wrfnat files? That may be the difference. Let me give that a check.

christinaholtNOAA commented:
Okay, I've pushed all my changes. The last attempt at `make_ics` ran fine with the wrfprs files with HRRR ICS. 

@JeffBeck-NOAA If you don't mind giving it another quick test, that'd be great.

JeffBeck-NOAA commented:
Thanks, @christinaholtNOAA.  I retested all external model combinations using FV3GFS, HRRR, and RAP, and they are all staged correctly and work with chgres_cube.  Approving!

christinaholtNOAA commented:
That's great news, @JeffBeck-NOAA! Thanks!

@gsketefian Are you okay with the changes I made in response to your review?

gsketefian commented:
@christinaholtNOAA Thanks for your responses, I'm good with them.  One more thing though.  After looking through the PR comments, I did not see any mention of the tests that are for fetching files from HPSS.  There are in `regional_workflow/tests/WE2E/test_configs/wflow_features`.  They are:
```
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2019061200
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2019101818
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2020022518
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2020022600
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2021010100
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019101818
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2020022518
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2020022600
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2021010100
get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_netcdf_2021062000
get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS
get_from_HPSS_ics_HRRR_lbcs_RAP
get_from_HPSS_ics_RAP_lbcs_RAP
```
The various dates with grib2 and nemsio are there because those are transition dates (e.g. where the directory naming convention changes or something similar).  Do you mind running these as well just to be sure all is working (unless you ran them already and I missed it)?  They are all on the 25km CONUS grid and should go pretty quick.  Thanks.

christinaholtNOAA commented:
@gsketefian  Sure. I can run them.

christinaholtNOAA commented:
@gsketefian I ran the tests you requested. They required some changes to accommodate the situation where "atmos" is included in the path. I've pushed those changes and passed all those hpss tests on Hera. 

I am running the other 25km tests again to make sure these changes don't break anything there. I'd like to merge this once I can confirm those pass.

christinaholtNOAA commented:
I can now confirm that all the 25 km tests we've discussed above pass on Hera.

gsketefian commented:
@christinaholtNOAA Great!  Let's merge it.
